File: backend/src/main/java/br/ufpr/sept/so2/modules/academico/infrastructure/persistence/DisciplinaJpaEntity.java
package br.ufpr.sept.so2.modules.academico.infrastructure.persistence;

import br.ufpr.sept.so2.modules.academico.domain.Disciplina;
import br.ufpr.sept.so2.shared.infrastructure.persistence.BaseEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;

import java.util.UUID;

@Entity
@Table(name = "disciplina")
public class DisciplinaJpaEntity extends BaseEntity {

    @Column(name = "id_curso", nullable = false)
    private UUID idCurso;

    @Column(nullable = false, length = 20)
    private String codigo;

    @Column(nullable = false, length = 200)
    private String nome;

    @Column(nullable = false)
    private int periodo;

    @Column(name = "carga_horaria_total", nullable = false)
    private int cargaHorariaTotal;

    @Column(nullable = false)
    private int creditos;

    @Column(nullable = false)
    private boolean ativa;

    protected DisciplinaJpaEntity() {
    }

    public static DisciplinaJpaEntity fromDomain(Disciplina disciplina) {
        DisciplinaJpaEntity entity = new DisciplinaJpaEntity();
        entity.setId(disciplina.getId());
        entity.idCurso = disciplina.getIdCurso();
        entity.codigo = disciplina.getCodigo();
        entity.nome = disciplina.getNome();
        entity.periodo = disciplina.getPeriodo();
        entity.cargaHorariaTotal = disciplina.getCargaHorariaTotal();
        entity.creditos = disciplina.getCreditos();
        entity.ativa = disciplina.isAtiva();
        return entity;
    }

    public void merge(Disciplina disciplina) {
        this.codigo = disciplina.getCodigo();
        this.nome = disciplina.getNome();
        this.periodo = disciplina.getPeriodo();
        this.cargaHorariaTotal = disciplina.getCargaHorariaTotal();
        this.creditos = disciplina.getCreditos();
        this.ativa = disciplina.isAtiva();
    }

    public Disciplina toDomain() {
        return new Disciplina(
                getId(),
                idCurso,
                codigo,
                nome,
                periodo,
                cargaHorariaTotal,
                creditos,
                ativa,
                getCreatedAt(),
                getUpdatedAt()
        );
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/academico/infrastructure/persistence/DisciplinaJpaEntity.kt
package br.ufpr.sept.so2.modules.academico.infrastructure.persistence

import br.ufpr.sept.so2.modules.academico.domain.Disciplina
import br.ufpr.sept.so2.shared.infrastructure.persistence.BaseEntity
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Table
import java.util.UUID

@Entity
@Table(name = "disciplina")
class DisciplinaJpaEntity : BaseEntity() {
    @Column(name = "id_curso", nullable = false)
    var idCurso: UUID? = null

    @Column(nullable = false, length = 20)
    var codigo: String = ""

    @Column(nullable = false, length = 200)
    var nome: String = ""

    @Column(nullable = false)
    var periodo: Int = 0

    @Column(name = "carga_horaria_total", nullable = false)
    var cargaHorariaTotal: Int = 0

    @Column(nullable = false)
    var creditos: Int = 0

    @Column(nullable = false)
    var ativa: Boolean = false

    fun merge(disciplina: Disciplina) {
        codigo = disciplina.codigo
        nome = disciplina.nome
        periodo = disciplina.periodo
        cargaHorariaTotal = disciplina.cargaHorariaTotal
        creditos = disciplina.creditos
        ativa = disciplina.ativa
    }

    fun toDomain() = Disciplina(
        requireNotNull(id),
        requireNotNull(idCurso),
        codigo,
        nome,
        periodo,
        cargaHorariaTotal,
        creditos,
        ativa,
        requireNotNull(createdAt),
        requireNotNull(updatedAt),
    )

    companion object {
        fun fromDomain(disciplina: Disciplina) = DisciplinaJpaEntity().apply {
            id = disciplina.id
            idCurso = disciplina.idCurso
            codigo = disciplina.codigo
            nome = disciplina.nome
            periodo = disciplina.periodo
            cargaHorariaTotal = disciplina.cargaHorariaTotal
            creditos = disciplina.creditos
            ativa = disciplina.ativa
        }
    }
}
